x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/aggregation/IrateDoubleAggregator.java

```diff
@@ -0,0 +1,214 @@
+
+    private static class DoubleIrateState {
+        static final long BASE_RAM_USAGE = RamUsageEstimator.sizeOfObject(DoubleIrateState.class);
+        long lastTimestamp;
+        long secondLastTimestamp = -1;
+        double lastValue;
+        double secondLastValue;
+        boolean hasSecond;
+
+        DoubleIrateState(long lastTimestamp, double lastValue) {
+            this.lastTimestamp = lastTimestamp;
+            this.lastValue = lastValue;
+            this.hasSecond = false;
+        }
+
+        long bytesUsed() {
+            return BASE_RAM_USAGE;
+        }
+    }
+
+    public static final class DoubleIrateGroupingState implements Releasable, Accountable, GroupingAggregatorState {
+        private ObjectArray<DoubleIrateState> states;
+        private final BigArrays bigArrays;
+        private final CircuitBreaker breaker;
+        private long stateBytes; // for individual states
+
+        DoubleIrateGroupingState(BigArrays bigArrays, CircuitBreaker breaker) {
+            this.bigArrays = bigArrays;
+            this.breaker = breaker;
+            this.states = bigArrays.newObjectArray(1);
+        }
+
+        void ensureCapacity(int groupId) {
+            states = bigArrays.grow(states, groupId + 1);
+        }
+
+        void adjustBreaker(long bytes) {
+            breaker.addEstimateBytesAndMaybeBreak(bytes, "<<rate aggregation>>");
+            stateBytes += bytes;
+            assert stateBytes >= 0 : stateBytes;
+        }
+
+        void append(int groupId, long timestamp, double value) {
+            var state = states.get(groupId);
+            if (state == null) {
+                state = new DoubleIrateState(timestamp, value);
+                states.set(groupId, state);
+                adjustBreaker(state.bytesUsed());
+            } else {
+                // We only need the last two values, but we need to keep them sorted by timestamp.
+                if (timestamp > state.lastTimestamp) {
+                    // new timestamp is the most recent
+                    state.secondLastTimestamp = state.lastTimestamp;
+                    state.secondLastValue = state.lastValue;
+                    state.lastTimestamp = timestamp;
+                    state.lastValue = value;
+                    state.hasSecond = true;
+                } else if (timestamp > state.secondLastTimestamp) {
+                    // new timestamp is the second most recent
+                    state.secondLastTimestamp = timestamp;
+                    state.secondLastValue = value;
+                    state.hasSecond = true;
+                } // else: ignore, too old
+            }
+        }
+
+        void combine(int groupId, LongBlock timestamps, DoubleBlock values, int otherPosition) {
+            final int valueCount = timestamps.getValueCount(otherPosition);
+            if (valueCount == 0) {
+                return;
+            }
+            final int firstIndex = timestamps.getFirstValueIndex(otherPosition);
```

Comment:
Most of the time, the firstIndex can be used for both blocks, but I think we need to use the first index for each block separately.